### PR TITLE
[`Core tokenization`]  `add_dummy_prefix_space` option to help with latest issues

### LINKED ISSUE #28622: Can `LlamaTokenizerFast` support the argument `add_prefix_space = False`
### System Info

With `transformers==4.36.2`

It seems the argument `add_prefix_space` is invalid here. 

### Who can help?

@ArthurZucker 

### Reproduction

```
>>> from transformers import LlamaTokenizerFast
>>> tokenizer = LlamaTokenizerFast.from_pretrained("hf-internal-testing/llama-tokenizer", add_prefix_space = False)
>>> tokenizer.tokenize("hello")
['▁hello']
>>> tokenizer.decode(tokenizer.encode("hello"))
'<s> hello'
```

### Expected behavior

Is there a bug, or is it my wrong usage?

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Allows users to use `tokenizer.tokenize` controlling the addition of prefix space. Let's also update fast! 

fixes #28622

### CODE REVIEW

File: src/transformers/convert_slow_tokenizer.py [hunk 1]
```diff
@@ -23,6 +23,7 @@
 from typing import Dict, List, Tuple
 
 from packaging import version
+
```
Comment (ArthurZucker) [inline src/transformers/convert_slow_tokenizer.py:26]:
```suggestion
```

**Comment by gabegrand** [comment]
Just wanted to say this would be hugely helpful for us over at https://github.com/probcomp/hfppl !

**Comment by haileyschoelkopf** [comment]
Likewise the ability to not include an extra SPIECE_UNDERLINE / Llama token 29871 when encoding a word with a space in front ( ` <word>`) would be huge for https://github.com/EleutherAI/lm-evaluation-harness !

**Comment by ArthurZucker** [comment]
will make it for next release I hope!

File: src/transformers/models/llama/tokenization_llama.py [hunk 2]
```diff
@@ -245,7 +250,11 @@ def tokenize(self, text: "TextInput", add_special_tokens=False, **kwargs) -> Lis
         if self.legacy or len(text) == 0:
             return super().tokenize(text, **kwargs)
 
-        tokens = super().tokenize(SPIECE_UNDERLINE + text.replace(SPIECE_UNDERLINE, " "), **kwargs)
+        text = text.replace(SPIECE_UNDERLINE, " ")
+        if kwargs.pop("add_prefix_space", self.add_prefix_space):
+            text = SPIECE_UNDERLINE + text
```
Comment (DNGros) [inline src/transformers/models/llama/tokenization_llama.py:255]: do you want to do these additions before you do the replace?

**Comment by claudiosv** [comment]
Would love to see this as well! Thanks for working on this.

File: src/transformers/models/llama/tokenization_llama.py (hunk 2 above)
Comment (ArthurZucker) [reply src/transformers/models/llama/tokenization_llama.py:255]: doesn't make a different I think

File: src/transformers/models/llama/tokenization_llama_fast.py [hunk 3]
```diff
@@ -100,6 +100,10 @@ class LlamaTokenizerFast(PreTrainedTokenizerFast):
             Whether or not to add an `eos_token` at the end of sequences.
         use_default_system_prompt (`bool`, *optional*, defaults to `False`):
             Whether or not the default system prompt for Llama should be used.
+        add_prefix_space (`bool`, *optional*):
+            Whether or not the tokenizer should automatically add a prefix space
+        from_slow (`book`, *optional*, defaults to `False`):
+            Whether or not the tokenizer should be converted from a slow one. If `add_prefix_space` is set, this will be set to `True`.
```
Comment (ArthurZucker) [inline src/transformers/models/llama/tokenization_llama_fast.py:106]:
```suggestion
```

File: src/transformers/models/llama/tokenization_llama_fast.py [hunk 4]
```diff
@@ -119,8 +123,16 @@ def __init__(
         add_bos_token=True,
         add_eos_token=False,
         use_default_system_prompt=False,
+        add_prefix_space=None,
+        from_slow=False,
```
Comment (ArthurZucker) [inline src/transformers/models/llama/tokenization_llama_fast.py:127]:
```suggestion
```

File: src/transformers/models/llama/tokenization_llama_fast.py [hunk 5]
```diff
@@ -119,8 +123,16 @@ def __init__(
...
         **kwargs,
     ):
+        if add_prefix_space is not None:
+            logger.warning_once(
+                "You set `add_prefix_space`. The tokenizer needs to be converted from the slow tokenizers"
+            )
+            from_slow = True
```
Comment (ArthurZucker) [inline src/transformers/models/llama/tokenization_llama_fast.py:134]:
```suggestion
            kwargs["from_slow"] = True
```

File: src/transformers/models/llama/tokenization_llama_fast.py [hunk 6]
```diff
@@ -131,6 +143,7 @@ def __init__(
             add_bos_token=add_bos_token,
             add_eos_token=add_eos_token,
             use_default_system_prompt=use_default_system_prompt,
+            from_slow=from_slow,
```
Comment (ArthurZucker) [inline src/transformers/models/llama/tokenization_llama_fast.py:146]:
```suggestion
```

**Review by ArthurZucker — COMMENTED** [review]
I'll let @lysandre decide, but instead of following what we do with bloom I'd rather we convert from slow. Bit slower but at least we are sure we use the correct logic. 
This is done with a warning.

**Comment by ArthurZucker** [comment]
Failing test is unrelated 😉

**Review by LysandreJik — APPROVED** [review]
Ok, this looks good to me

File: src/transformers/models/llama/tokenization_llama_fast.py [hunk 7]
```diff
@@ -119,8 +121,15 @@ def __init__(
...
         add_eos_token=False,
         use_default_system_prompt=False,
+        add_prefix_space=None,
         **kwargs,
     ):
+        if add_prefix_space is not None:
+            logger.warning_once(
+                "You set `add_prefix_space`. The tokenizer needs to be converted from the slow tokenizers"
+            )
+            kwargs["from_slow"] = True
```
Comment (LysandreJik) [inline src/transformers/models/llama/tokenization_llama_fast.py:131]: How long does it take to convert the tokenizer from slow? If it's quick we can move it to `info`

Comment (ArthurZucker) [reply src/transformers/models/llama/tokenization_llama_fast.py:131]: Around 10 seconds I believe!

**Comment by casper-hansen** [comment]
@ArthurZucker I know this is old at this point, but this PR seems to introduce a unintended side-effect for Mistral v0.1 tokenizers. I figured out that I can fix it by always initializing with `add_prefix_space=True`. Not sure if this is on your radar or if it is fixed in newer versions, but would totally appreciate if this can be fixed for the Mistral model series.

Tokens before (4.37.2):

```
SLICED: ['1: <s>', '330: A', '28747: :', '28705: ', '28740: 1', '28783: 8', '13: <0x0A>']
```

Tokens after (this PR):

```
SLICED: ['1: <s>', '28747: :', '28705:  ', '28740: 1', '28783: 8', '13: <0x0A>']
```

image

https://www.diffchecker.com/GwA54pMf/

**Pull request merged**